### PR TITLE
ThisKeyword is not a type, only ThisType is

### PR DESCRIPTION
**Pull request by andy-ms**
Replaces #15686
Ensures that we're always using `ThisType` for `this` in a type position, not `ThisKeyword`.

### CODE REVIEW

File: tests/cases/fourslash/quickInfoOnThis.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@
 ////}
 
 verify.quickInfos({
-    0: "this: this",
+    0: "class Foo",
```
Comment (andy-ms) [inline tests/cases/fourslash/quickInfoOnThis.ts]: This is in type position, not an expression, so I think this change is correct.

File: src/compiler/scanner.ts [hunk 2]
```diff
@@ -285,9 +285,13 @@ namespace ts {
     }
 
     const tokenStrings = makeReverseMap(textToToken);
+    // "this" maps to SyntaxKind.ThisKeyword normally, so can't be in textToToken. But can put it in tokenStrings.
+    tokenStrings[SyntaxKind.ThisType] = "this";
 
     export function tokenToString(t: SyntaxKind): string | undefined {
-        return tokenStrings[t];
+        const res = tokenStrings[t];
+        Debug.assert(!!res);
```
Comment (ajafff) [inline src/compiler/scanner.ts]: That's a breaking change since it used to work with any SyntaxKind before, e.g. `StringLiteral` which is a token and should therefore be a valid input. 

Otherwise you can change the return type and remove the undefined there.

Comment (andy-ms) [reply src/compiler/scanner.ts]: Oh well, better to avoid the breaking change.

File: src/compiler/scanner.ts [hunk 3]
```diff
@@ -285,6 +285,8 @@ namespace ts {
     }
 
     const tokenStrings = makeReverseMap(textToToken);
+    // "this" maps to SyntaxKind.ThisKeyword normally, so can't be in textToToken. But can put it in tokenStrings.
+    tokenStrings[SyntaxKind.ThisType] = "this";
```
Comment (rbuckton) [inline src/compiler/scanner.ts]: Why is this needed?

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -5012,7 +5012,7 @@ namespace ts {
             || kind === SyntaxKind.BooleanKeyword
             || kind === SyntaxKind.StringKeyword
             || kind === SyntaxKind.SymbolKeyword
-            || kind === SyntaxKind.ThisKeyword
+            || kind === SyntaxKind.ThisType
```
Comment (rbuckton) [inline src/compiler/utilities.ts]: `ThisType` is already within the range of `FirstTypeNode` through `LastTypeNode`.

File: src/services/completions.ts [hunk 5]
```diff
@@ -1674,6 +1674,10 @@ namespace ts.Completions {
         function getAllKeywordCompletions() {
             const allKeywordsCompletions: CompletionEntry[] = [];
             for (let i = SyntaxKind.FirstKeyword; i <= SyntaxKind.LastKeyword; i++) {
+                if (i === SyntaxKind.ThisType) {
```
Comment (rbuckton) [inline src/services/completions.ts]: `ThisType` is not a keyword.

File: src/compiler/types.ts [hunk 6]
```diff
@@ -157,6 +157,7 @@ namespace ts {
         SuperKeyword,
         SwitchKeyword,
         ThisKeyword,
+        ThisType, // Same as ThisKeyword, but parsed from a type context.
```
Comment (rbuckton) [inline src/compiler/types.ts]: A `ThisType` is not a keyword. 👎

**Comment by rbuckton** [comment]
The reason we have a `ThisType` vs. just using `ThisKeyword` has to do with transformations and lexical `this` capturing. Without distinguishing between the two, `computeTransformFlagsForNode` can't distinguish between a lexical `this` and a `this` in a type position.

File: src/compiler/types.ts (hunk 6 above)
Comment (andy-ms) [reply src/compiler/types.ts]: Should it contain a keyword as a child? Or should it be a leaf node, but still not a keyword?

**Comment by mhegazy** [comment]
closing for now.
